juju/client/overrides.py

```diff
@@ -273,3 +274,26 @@ def __repr__(self):
         return '<{} source={} value={}>'.format(type(self).__name__,
                                                 repr(self.source),
                                                 repr(self.value))
+
+class Resource(Type):
+    _toSchema = {'application': 'application', 'charmresource': 'CharmResource', 'id_': 'id', 'pending_id': 'pending-id', 'timestamp': 'timestamp', 'username': 'username', 'name': 'name', 'origin': 'origin'}
+    _toPy = {'CharmResource': 'charmresource', 'application': 'application', 'id': 'id_', 'pending-id': 'pending_id', 'timestamp': 'timestamp', 'username': 'username', 'name': 'name', 'origin': 'origin'}
+    def __init__(self, charmresource=None, application=None, id_=None, pending_id=None, timestamp=None, username=None, name=None, origin=None, **unknown_fields):
+        '''
+        charmresource : CharmResource
+        application : str
+        id_ : str
+        pending_id : str
+        timestamp : str
+        username : str
+        name: str
+        origin : str
```

Comment:
Why does this need to be in overrides?  Are these new fields from more recent builds of juju?  If that's the case, we should go ahead and follow the [upstream updates](https://pythonlibjuju.readthedocs.io/en/latest/upstream-updates/index.html) process to get all of the changes.  If these fields are not new, I'd like to figure out why they weren't captured already.

Reply:
These aren't new, I've observed the same issue with Juju controllers running 2.0.2 and 2.2.3.

The schema is wrong. In particular, this bit of the Resource definition states that there's a field named CharmResource, defined by #/definitions/CharmResource:

```
"Resource": {
  "type": "object",
  "properties": {
    "CharmResource": {
      "$ref": "#/definitions/CharmResource"
    },
    ...
  }
  ...
}
```

where #/definitions/CharmResource defines the `name` and `origin` fields:

```
"CharmResource": {
  "type": "object",
  "properties": {
    ...
    "name": {
      "type": "string"
    },
    "origin": {
      "type": "string" 
    },
    ...
  },
  ...
},
```

However, here's the actual data I get in the response:

```
"resources": [
  {
    "username": "test/0",
    "type": "file",
    "pending-id": "",
    "name": "test-new",
    "fingerprint": "XwyivFD/IYBYXbtVo2VwSITDq8/yv+tncsHIPTnEY46+DjYpyL2JOupZXFTjSkz7",
    "revision": 3,
    "size": 10,
    "id": "test/test-new",
    "timestamp": "2017-09-27T15:07:53.923Z",
    "path": "test",
    "origin": "store",
    "application": "test"
  }
]
```

So Resource does not have a field named CharmResource. Rather, Resource _includes the fields from the CharmResource definition_. But that's not what the schema says.

After reading the [JSON Schema doc page that discusses $ref](https://spacetelescope.github.io/understanding-json-schema/structuring.html), it looks to me like the Resource definition should be using `allOf` to extend from CharmResource properly.

@johnsca Seems like I should open a bug against juju/schemagen, yeah?

Reply:
That seems to be how it's [defined](https://github.com/juju/juju/blob/develop/apiserver/params/resources.go#L78) in the juju source, though I guess the lack of an explicit `` `json:"<name>"` `` is supposed to make it be included inline.  We're still using a [fork](https://github.com/bcsaller/jsonschema) of the go [jsonschema](https://github.com/alecthomas/jsonschema) library which actually does the reflection, so this might already fixed upstream.  I created juju/schemagen#1 to track getting us off the fork or at least getting it updated.